packages/scope-manager/src/referencer/Referencer.ts

```diff
@@ -707,7 +695,13 @@ class Referencer extends Visitor {
   }
 
   protected TSTypeAssertion(node: TSESTree.TSTypeAssertion): void {
-    this.visitTypeAssertion(node);
+    if (node.kind === 'as') {
+      this.visit(node.expression);
+      this.visitType(node.typeAnnotation);
+    } else {
+      this.visitType(node.typeAnnotation);
+      this.visit(node.expression);
+    }
```

Comment:
while doing this change i noticed that order of visiting expression and typeAssertion was not correct